contracts/tezos/tezosSwap.py

```diff
@@ -0,0 +1,420 @@
+            self.data.pairs[pair] = _pairs[pair]
+
+    """
+        Remove swap pair
+
+        args:
+            _pairs: A list of pair names
+    """
+
+    @sp.entry_point
+    def removeSwapPair(self, _pairs):
+        self.onlyByAdmin()
+        sp.for pair in _pairs:
+            del self.data.pairs[pair]
+
+    """
+        Update swap fees
+
+        args:
+            _pairs: A list of pair names
+    """
+
+    @sp.entry_point
+    def updateSwapFees(self, _swapFee, _commission):
+        self.onlyByAdmin()
+        self.data.commission=_commission
+        self.data.swapFee=_swapFee
+
+    """
+        Sets baker for the contract
+
+        args:
+            _baker: A baker address
+    """
+
+    @sp.entry_point
+    def setBaker(self, _baker):
+        self.onlyByAdmin()
+        sp.set_delegate(_baker)
+
+    """
+        Transfers the accumulated service fee
+
+        args:
+            _to: Account address to transfer the fees
+    """
+
+    @sp.entry_point
+    def transferFees(self, _to):
+        self.onlyByAdmin()
+        sp.verify((sp.balance-sp.mutez(self.data.userXTZDeposit))>sp.mutez(0))
+        sp.send(_to, (sp.balance-sp.mutez(self.data.userXTZDeposit)))
+
+    """
+        Initiate new swap
+
+        args:
+            _swapHash: swap hash
+            _value: amount of tokens exchanged in the swap
+            _expected: amount of tokens expected in return
+            _refundTimestamp: unix time(sec) after which the swap expires
+            _pair: The swap pair being exchanged
+            _asset: Current token 
+    """
+
+    @sp.entry_point
+    def offer(self, _swapHash, _value, _expected, _refundTimestamp, _pair, _asset):
+        self.isInitiable(_swapHash, _refundTimestamp, _pair, _asset)
+        sp.if _asset=="xtz":
+            sp.verify(sp.amount==sp.mutez(_value+self.data.commission),"AMOUNT_MISMATCH")
+            self.data.userXTZDeposit+= _value+self.data.commission
+        sp.else:
+            sp.verify(sp.amount==sp.mutez(self.data.commission), "AMOUNT_MISMATCH")
+            self.data.userXTZDeposit+= self.data.commission
+            c = sp.contract(TTransfer, self.data.pairs[_pair][_asset].contract, entry_point="transfer").open_some()
+            transferData = sp.record(from_=sp.sender, to_=sp.self_address, value=_value)
+            sp.transfer(transferData, sp.mutez(0), c)
+
+        self.data.swaps[_swapHash] = sp.record(swapHash=_swapHash, initiator=sp.sender, refundTimestamp=_refundTimestamp, value=_value, expected=_expected, asset=_asset,pair=_pair, commission=self.data.commission)
+
+    """
+        Redeem the swap if possible
+
+        args:
+            _swapHash: swap hash
+            _value: amount of tokens exchanged in the swap, matches the expected amount for the swap
+            _pair: The swap pair being exchanged
+            _asset: Current token asset 
+    """
+
+    @sp.entry_point
+    def take(self, _swapHash, _value, _pair, _asset):
```

Comment:
why not read everything except `_swapHash` from storage, like it does in `refund()`

Reply:
_asset is different for the take call, having these parameters makes sure that a wrong call (for different asset and pair) was not made